ofrak_core/ofrak/core/gzip.py

```diff
@@ -41,44 +39,43 @@ class GzipUnpacker(Unpacker[None]):
-                return await resource.create_child(
-                    tags=(GenericBinary,),
-                    data=gzip_file.read(),
-                )
-        except BadGzipFile:
-            # Create temporary file with .gz extension
-            with tempfile.NamedTemporaryFile(suffix=".gz") as temp_file:
-                temp_file.write(data)
-                temp_file.flush()
-                cmd = [
-                    "pigz",
-                    "-d",
-                    "-c",
-                    temp_file.name,
-                ]
-                proc = await asyncio.create_subprocess_exec(
-                    *cmd,
-                    stdout=asyncio.subprocess.PIPE,
-                    stderr=asyncio.subprocess.PIPE,
-                )
-                stdout, stderr = await proc.communicate()
-                data = stdout
-                if proc.returncode:
-                    # Forward any gzip warning message and continue
-                    if proc.returncode == -2 or proc.returncode == 2:
-                        LOGGER.warning(stderr)
-                        data = stdout
-                    else:
-                        raise CalledProcessError(returncode=proc.returncode, cmd=cmd, stderr=stderr)
-
-                await resource.create_child(
-                    tags=(GenericBinary,),
-                    data=data,
-                )
+        if len(data) < 1024 * 1024 * 4:  # Use pigz for more than 4MiB
+            # wbits > 16 handles the gzip header and footer
+            # We need to create a zlib.Decompress object for Python < 3.11
+            # to set this parameter
+            chunks = []
+
+            decompressor = zlib.decompressobj(wbits=16 + zlib.MAX_WBITS)
+            while data.startswith(b"\037\213"):
+                chunks.append(decompressor.decompress(data))
+                if decompressor.eof:
+                    break
+                data = decompressor.unused_data.lstrip(b"\0")
+
+            if not len(chunks):
+                raise ValueError("Not a gzipped file")
+
+            return await resource.create_child(tags=(GenericBinary,), data=b"".join(chunks))
+        else:
+            cmd = ["pigz", "-d"]
+            proc = await asyncio.create_subprocess_exec(
+                *cmd,
+                stdin=asyncio.subprocess.PIPE,
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+            )
+            stdout, stderr = await proc.communicate(data)
+            if proc.returncode:
+                # Forward any gzip warning message and continue
+                if proc.returncode == -2 or proc.returncode == 2:
+                    LOGGER.warning(stderr)
+                else:
+                    raise CalledProcessError(returncode=proc.returncode, cmd=cmd, stderr=stderr)
+
+            await resource.create_child(
+                tags=(GenericBinary,),
+                data=stdout,
+            )
```

Comment:
Probably makes sense to break each case of this `if` into its own helper function. 

The reason is that we mandate 100% function coverage in testing before code can be merged, but we have no requirements about line coverage or branch coverage. We want to make sure that we test both of these cases, with no possibility of regressions in the future. As such, if each case is its own function, our function coverage testing will confirm that both branches are hit. 

I suspect that all of our test files are small, and that the `pigz` case is never tested.

Reply:
Refactored into separate functions and added test case with a big binary for coverage